herd/BPFSem.ml

```diff
@@ -0,0 +1,249 @@
+(*                                                                          *)
+(* This software is governed by the CeCILL-B license under French law and   *)
+(* abiding by the rules of distribution of free software. You can use,      *)
+(* modify and/ or redistribute the software under the terms of the CeCILL-B *)
+(* license as circulated by CEA, CNRS and INRIA at the following URL        *)
+(* "http://www.cecill.info". We also give a copy in LICENSE.txt.            *)
+(****************************************************************************)
+
+(** Semantics of BPF instructions *)
+
+module Make (C : Sem.Config) (V : Value.S with type Cst.Instr.t = BPFBase.instruction) =
+struct
+  module BPF = BPFArch_herd.Make (SemExtra.ConfigToArchConfig (C)) (V)
+  module Act = MachAction.Make (C.PC) (BPF)
+  include SemExtra.Make (C) (BPF) (Act)
+
+  (* Barrier pretty print *)
+  let sync = None
+  let barriers = []
+  let isync = None
+
+  (*  TODO: let nat_sz = MachSize.Quad (* 64-bit Registers *) *)
+  let nat_sz = V.Cst.Scalar.machsize
+  let atomic_pair_allowed _ _ = true
+
+  (********************)
+  (* Semantics proper *)
+  (********************)
+
+  module Mixed (SZ : ByteSize.S) = struct
+    let ( >>= ) = M.( >>= )
+    let ( >>*= ) = M.( >>*= )
+    let ( >>| ) = M.( >>| )
+    let ( >>! ) = M.( >>! )
+    let ( >>:: ) = M.( >>:: )
+    let unimplemented op = Warn.user_error "BPF operation %s is not implemented (yet)" op
+
+    let tr_opamo op =
+      match op with
+      | BPF.AMOXCHG -> assert false
+      | BPF.ADD -> Op.Add
+      | BPF.AND -> Op.And
+      | BPF.OR -> Op.Or
+      | BPF.XOR -> Op.Xor
+      | BPF.AMOCMPXCHG -> assert false
+      | _ -> unimplemented "atomic op"
+    ;;
+
+    let tr_op = function
+      | BPF.ADD -> Op.Add
+      | BPF.SUB -> Op.Sub
+      | BPF.AND -> Op.And
+      | BPF.OR -> Op.Or
+      | BPF.XOR -> Op.Xor
+      | BPF.MUL -> Op.Mul
+      | BPF.DIV -> Op.Div
+      | BPF.REM -> Op.Rem
+      | BPF.LSL -> Op.ShiftLeft
+      | BPF.LSR -> Op.Lsr
+      | BPF.ASR -> unimplemented (BPF.pp_op BPF.ASR)
+      | BPF.AMOCMPXCHG -> unimplemented "non-atomic CMPXCHG"
+      | BPF.AMOXCHG -> unimplemented "non-atomic XCHG"
+    ;;
+
+    let tr_cond cond =
+      match cond with
+      | BPF.EQ -> Op.Eq
+      | BPF.NE -> Op.Ne
+      | BPF.LT -> Op.Lt
+      | BPF.GE -> Op.Ge
+    ;;
+
+    let mk_read sz ato loc v =
+      Act.Access (Dir.R, loc, v, ato, (), sz, Act.access_of_location_std loc)
+    ;;
+
+    let read_reg is_data r ii =
+      M.read_loc is_data (mk_read nat_sz BPF.N) (A.Location_reg (ii.A.proc, r)) ii
+    ;;
+
+    let read_reg_ord = read_reg false
+    let read_reg_data = read_reg true
+
+    let do_read_mem sz ato a ii =
+      M.read_loc false (mk_read sz ato) (A.Location_global a) ii
+    ;;
+
+    let read_mem sz a ii = do_read_mem sz BPF.N a ii
+    let read_mem_atomic sz a ii = do_read_mem sz BPF.X a ii
```

Comment:
Function `read_mem_atomic` is not called.

Reply:
Fixed in next version